scripts/listener.js

```diff
@@ -0,0 +1,302 @@
+const process = require('process')
+const http = require('http')
+const urllib = require('url')
+const Origin = require('../dist/index') // Will eventualy be the origin npm package
+const Web3 = require('web3')
+
+const web3Provider = new Web3.providers.HttpProvider('http://localhost:8545')
+const web3 = new Web3(web3Provider)
+const o = new Origin({ web3 })
+
+// Origin Listener
+// ---------------
+// An at-least-once event listener for origin events (May deliver an event more than once).
+// Make sure your webhook endpoint is idempotent!
+// Sends events along with all the related data.
+// Designed to be infura compatible - API calls only, no subscriptions.
+//
+// To use:
+// - Run `npm start run` to setup a local IPFS and blockchain, and run tests
+// - In another terminal run `node scripts/listener.js`
+// - Or `node scripts/listener.js --webhook=http://localhost/originevents/`
+//
+// Todo
+// - Allow configuring web3 endpoint and IPFS endpoint
+// - Handle errors
+// - Persist starting point
+// - Handle blockchain splits/winners
+// - Include current-as-of block numbers in POSTs
+// - Perhaps send related data as it was at the time of the event, not crawl time
+
+// -----------------------------
+// Section 1: Follow rules
+// -----------------------------
+
+// Helper functions
+const listingId = log => {
+  return [log.networkId, log.contractVersionKey, log.decoded.listingID].join(
+    '-'
+  )
+}
+const offerId = log => {
+  return [
+    log.networkId,
+    log.contractVersionKey,
+    log.decoded.listingID,
+    log.decoded.offerID
+  ].join('-')
+}
+const getListingDetails = async log => {
+  return {
+    listing: await o.marketplace.getListing(listingId(log))
+  }
+}
+const getOfferDetails = async log => {
+  return {
+    listing: await o.marketplace.getListing(listingId(log)),
+    offer: await o.marketplace.getOffer(offerId(log))
+  }
+}
+
+// Rules for acting on events
+// Adding a rule here makes the listener listen for the event.
+const LISTEN_RULES = {
+  v00_MarketplaceContract: {
+    ListingCreated: getListingDetails,
```

Comment:
I've got them ordered as they appear in the contract. This seems okay, since that is ordered by type (listing, then offer), then ordered by when it appears in the contract.

Reply:
Sounds good.